// @jswork/next-content-type/dist/index.js
/*!
 * name: @jswork/next-content-type
 * description: Get correct content type for next.
 * homepage: https://js.work
 * version: 1.0.8
 * date: 2023-04-20 22:42:46
 * license: MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports['default'] = void 0;
var _next = _interopRequireDefault(require('@jswork/next'));
function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}
var TYPES = {
  urlencoded: 'application/x-www-form-urlencoded',
  multipart: 'multipart/form-data',
  json: 'application/json;charset=utf-8',
  raw: 'text/plain',
  text: 'text/plain',
  blob: 'application/octet-stream',
  arraybuffer: 'application/octet-stream',
  stream: 'application/octet-stream',
  document: 'application/xml',
  auto: null
};
_next['default'].contentType = function (inKey) {
  return TYPES[inKey] || TYPES.urlencoded;
};
if (typeof module !== 'undefined' && module.exports) {
  module.exports = _next['default'].contentType;
}
var _default = _next['default'].contentType;
exports['default'] = _default;
